packages/compat/src/resolver.ts

```diff
@@ -437,16 +437,16 @@ export default class CompatResolver implements Resolver {
 
   private tryComponent(path: string, from: string, withRuleLookup = true): Resolution | null {
     let parts = path.split('@');
+    let cache = PackageCache.shared('embroider-stage3');
     if (parts.length > 1 && parts[0].length > 0) {
-      let cache = PackageCache.shared('embroider-stage3');
       let packageName = parts[0];
       let renamed = this.params.adjustImportsOptions.renamePackages[packageName];
       if (renamed) {
         packageName = renamed;
       }
       return this._tryComponent(parts[1], from, withRuleLookup, cache.resolve(packageName, cache.ownerOfFile(from)!));
     } else {
-      return this._tryComponent(path, from, withRuleLookup, this.appPackage);
+      return this._tryComponent(path, from, withRuleLookup, cache.ownerOfFile(from)!);
```

Comment:
Thanks for working on this.

As I said [here](https://github.com/embroider-build/embroider/issues/800#issuecomment-832690721), it's not correct to use `cache.ownerOfFile(from)` here. That's not how component resolution works -- they're not resolved relative to the current package, they're resolved relative to the containing engine.

The code that understands how to find the containing engine is [partitionEngines](https://github.com/embroider-build/embroider/blob/ea6f2ddce1878a8597bea708f984e46292ba743b/packages/core/src/app.ts#L551).